src/util/psbt/psbtmap/psbtglobal.rs

```diff
@@ -0,0 +1,130 @@
+use std::collections::HashMap;
+
+use blockdata::transaction::Transaction;
+use network::encodable::{ConsensusDecodable, ConsensusEncodable};
+use network::serialize::{deserialize, SimpleDecoder, SimpleEncoder};
+use util;
+use util::psbt::psbtmap::PSBTMap;
+use util::psbt::raw::{PSBTRawKVPair, PSBTRawKey};
+use util::psbt::serialize::PSBTSerialize;
+use util::psbt::Error;
+
+/// A key-value map for global data.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PSBTGlobal {
+    /// The unsigned transaction, scriptSigs and witnesses for each input must be
+    /// empty.
+    pub unsigned_tx: Transaction,
+    /// Unknown global key-value pairs.
+    pub unknown: HashMap<PSBTRawKey, Vec<u8>>,
+}
+
+impl PSBTMap for PSBTGlobal {
+    fn insert_pair(&mut self, pair: PSBTRawKVPair) -> Result<(), util::Error> {
+        let PSBTRawKVPair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(util::Error::from(Error::DuplicateKey(raw_key)));
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(util::Error::from(Error::DuplicateKey(raw_key)));
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<PSBTRawKVPair>, util::Error> {
+        let mut rv: Vec<PSBTRawKVPair> = Default::default();
+
+        rv.push(PSBTRawKVPair {
+            key: PSBTRawKey {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: self.unsigned_tx.psbt_serialize()?,
+        });
+
+        for (key, value) in self.unknown.iter() {
+            rv.push(PSBTRawKVPair {
+                key: key.clone(),
+                value: value.clone(),
+            });
+        }
+
+        Ok(rv)
+    }
+
+    fn merge(&mut self, other: Self) -> Result<(), util::Error> {
+        if self.unsigned_tx != other.unsigned_tx {
+            return Err(util::Error::ParseFailed);
+        }
+
+        self.unknown.extend(other.unknown);
+        Ok(())
+    }
+}
+
+impl From<Transaction> for Result<PSBTGlobal, util::Error> {
```

Comment:
It seems weird to implement this to a `std` type, isn't it more clear to have a `from_transaction` method in `PSBTGlobal`?

Something along the lines:

```rust
impl PSBTGlobal {
    pub fn from_transaction(tx: Transaction) -> Result<PSBTGlobal, util::Error> {
        ....
    }
}
```

Reply:
How about I add this and leave the `From` version as well?

Reply:
Done for both `PsbtGlobal` and `PartiallySignedTransaction`